feat: handle display name and namespace

fsuits opened the pull request:
## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Hot fix (emergency fix and release)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation (change which affects the documentation site)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Release (`develop` -> `main`)

## Quality assurance (all should be covered).

- [x] My code follows the code style of this project.
- [ ] Documentation for my change is up to date?
- [x] My PR meets testing requirements.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Summary
Handle display name as a special property in the markdown header
Allow specifying its namespace value with new default-namespace parameter
This only applies to profile-generate and assemble

closes #1161 
closes #1167

## Key links:

- [Sonar coverage](https://sonarcloud.io/dashboard?id=compliance-trestle)

# Before you merge

- Ensure it is a 'squash commit' if not a release.
- Ensure CI is currently passing
- Check sonar. If you are working for a fork a maintainer will reach out, if required.

enikonovad commented:
There is a questionable behavior of default-namespace, when trying to overwrite it.

### Scenario 1. Overwrite existing default-namespace in profile-generate and profile-assemble.
Start with profile with no ns in props (example current profile in test-profiles repo)
1. Run `author profile-generate --output md_profiles/SOMEPROFILE --name SOMEPROFILE -ns "https://some_ns"`
    1. Observe that markdowns generated with `default-namespace:https://some_ns` - Good
2. Run `author profile-assemble --markdown md_profiles/SOMEPROFILE --output SOMEPROFILE --set-parameters -ns "https://some_ns"`
    1. Observe that Json has `ns:https://some_ns` added to props - Good
3. Now we want to change default-namespace, run `author profile-generate --output md_profiles/SOMEPROFILE --name SOMEPROFILE -ns "https://ovewriten_ns"`
    1. **Expected**: `default-namespace: https://some_ns/` is changed to `default-namespace: https://ovewriten_ns/` in all markdowns 
    2. **Observed**: No changes to mds, i.e. `default-namespace: https://some_ns/` stayed

Continue with this profile.
4. Run `author profile-assemble --markdown md_profiles/SOMEPROFILE --output SOMEPROFILE --set-parameters -ns "https://ovewriten_ns"`
**Expected**: ns in props changed to “https://overwriten_ns/”
**Observed**: No changes to JSON    


Now no changes to markdown `default-namespace` in steps 3 seems like a bug to me. Step 4 behaviour is also seems odd.

enikonovad commented:
Also there is a possible artifact that I noticed. Profile-generate will add x-trestle-set-params for params that are not in `set-parameters` in Json profile. Maybe it has something to do with the fact that set-params in Json start with `prm_2` rather than `prm_1`.

Example Json:
```
"set-parameters": [
        {
          "param-id": "ac-1_prm_2",
          "props": [
            {
              "name": "display-name",
              "ns": "https://default_namespace",
              "value": "AC-1 (b) (1)"
            }
          ],
          "values": [
            "at least annually"
          ]
        },
        {
          "param-id": "ac-1_prm_3",
          "props": [
            {
              "name": "display-name",
              "ns": "http://some_url",
              "value": "AC-1 (b) (2)"
            }
          ],
          "values": [
            "at least annually"
          ]
        },
```

Markdown AC-1 will have three set-parameters
```
x-trestle-set-params:
  ac-1_prm_1: <-- Where did this came from?
    label: organization-defined personnel or roles
  ac-1_prm_2:
    profile-values: at least annually
    values:
    display-name: AC-1 (b) (1)
  ac-1_prm_3:
    profile-values: at least annually
    values:
    display-name: AC-1 (b) (2)
    ns: http://some_url
 ```
 
 Doing any changes to this "ghost" set-param in markdown will not change anything in Json

fsuits commented:
Thanks for doing these sanity checks - and those are all good questions.  I need to write documentation for it, in which case I hope it makes sense.

The intent is to allow specifying ns for things that don't have ns defined in the header.  Anything else will get that ns - and if something in the header has ns that matches the default - it will *not* be shown - to avoid cluttering the header.

The issue with ns not changing on second generate is that generating on top of existing contents is intended only to add in some sense - and that needs to be figured out better.  Any prose found in the md will not be altered by generate - only new sections will be added.  It's not clear why the second generate is happening and what we want it to do, but we want to avoid overwriting edits made to the markdown.  The user could have edited the value in the yaml header and it would have been changed by the generate operation.  I think we should do a high level review of this stuff and how it relates to the relevant use cases - so this is less confusing.

The second issue is due to the header capturing everything known about set-parameters in the profile.  What is shown in the generated header is every parameter as it exists in the resolved catalog - along with changes that are made by the set-parameters of this profile.  ac-1_prm_1 only has a label in the original catalog and there is no set-parameter for it in the profile.  So that is what is shown, and at this point a user could change the label and/or add values - whatever.  Then, if assembly is done with set-parameters True, the new set-parameters should appear along with their changes.

This behavior evolved in the work with Doug and I think this is an ok way to do things.  The goal is to be empowered to make additions to the profile during assembly, but careful with existing markdown when doing generate.

enikonovad commented:
Thanks for the detailed response, some further comments.

>  It's not clear why the second generate is happening and what we want it to do, but we want to avoid overwriting edits made to the markdown.

The intent here was to test a scenario if the user would like to change the default namespace. Say URL changed or something and now it needs to be updated. Since there is no way to change default namespace in the markdown (it is rightfully not printed), the only current way is to modify `default-namespace` in the header which will work but which user should not do. The cli option seems to be a second choice, but currently it will not overwrite things. The only other way is to delete md folder and regenerate from scratch. How we should go about it?

>Then, if assembly is done with set-parameters True, the new set-parameters should appear along with their changes.

I think this is where I was confused about it, since if I change the label of `ac-1_prm_1` from `    label: organization-defined personnel or roles` to `label: organization-defined time period` (or do any other changes to it, even deleting it entirely) and then assemble it using `author profile-assemble --markdown md_profiles/SOMEPROFILE --output SOMEPROFILE --set-parameters` it will not do anything to Json profile (i.e. it will not add set-param for ac-1_prm_1). Is this expected?

fsuits commented:
> Thanks for the detailed response, some further comments.
> 
> > It's not clear why the second generate is happening and what we want it to do, but we want to avoid overwriting edits made to the markdown.
> 
> The intent here was to test a scenario if the user would like to change the default namespace. Say URL changed or something and now it needs to be updated. Since there is no way to change default namespace in the markdown (it is rightfully not printed), the only current way is to modify `default-namespace` in the header which will work but which user should not do. The cli option seems to be a second choice, but currently it will not overwrite things. The only other way is to delete md folder and regenerate from scratch. How we should go about it?
> 
> > Then, if assembly is done with set-parameters True, the new set-parameters should appear along with their changes.
> 
> I think this is where I was confused about it, since if I change the label of `ac-1_prm_1` from ` label: organization-defined personnel or roles` to `label: organization-defined time period` (or do any other changes to it, even deleting it entirely) and then assemble it using `author profile-assemble --markdown md_profiles/SOMEPROFILE --output SOMEPROFILE --set-parameters` it will not do anything to Json profile (i.e. it will not add set-param for ac-1_prm_1). Is this expected?

I didn't express myself well for the first part.  I know why you did the second profile generate, but I don't know what the main use case need is for doing it a second time.  I believe the current approach is that it is able to add new guidance sections, but it cannot alter existing prose.  I need to clarify what is allowed to happen in the yaml header.  The yaml header may be overwritten by contents of a yaml header being handed in - but that is different from new set-parameters coming in.

For the second part, the documented behavior is only to change set-parameters on assemble *if* new values have been assigned.  I don't think there is a big need to change only a label value - but providing a value along with changing the label makese sense.

I will try to make a list of these ambiguities as I review the new documentation needs - and make changes for things that seem problematic.

fsuits commented:
Addressed the PR feedback per discussion.  default-namespace will now change in the header if requested on each profile-generate.  namespace for properties is handled more cleanly.  More tests added.

enikonovad reviewed (commented):
A small comment

fsuits commented on tests/test_utils.py:
```diff
@@ -186,17 +186,17 @@ def text_files_equal(path_a: pathlib.Path, path_b: pathlib.Path) -> bool:
     if not file_path.exists():
         raise TrestleError(f'Test file {file_path} not found.')
     lines: List[str] = []
     with file_path.open('r') as f:
```
added.  this is a file used for testing so the encoding is less critical.  There are several other files for testing that don't have encoding specified so we should be aware.  I changed all in this file.

enikonovad reviewed (approved):
LGTM
